src/lattice.cpp

```diff
@@ -0,0 +1,945 @@
+#include "lattice.h"
+
+#include <cmath>
+#include <sstream>
+#include <vector>
+
+#include "cell.h"
+#include "error.h"
+#include "geometry_aux.h"
+#include "hdf5_interface.h"
+#include "xml_interface.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Global variables
+//==============================================================================
+
+std::vector<Lattice*> lattices_c;
+
+std::unordered_map<int32_t, int32_t> lattice_dict;
+
+//==============================================================================
+// Lattice implementation
+//==============================================================================
+
+Lattice::Lattice(pugi::xml_node lat_node)
+{
+  if (check_for_node(lat_node, "id")) {
+    id = stoi(get_node_value(lat_node, "id"));
+  } else {
+    fatal_error("Must specify id of lattice in geometry XML file.");
+  }
+
+  if (check_for_node(lat_node, "name")) {
+    name = get_node_value(lat_node, "name");
+  }
+
+  if (check_for_node(lat_node, "outer")) {
+    outer = stoi(get_node_value(lat_node, "outer"));
+  }
+}
+
+//==============================================================================
+
+LatticeIter Lattice::begin()
+{return LatticeIter(*this, 0);}
```

Comment:
Thinking about this a little, I don't think you actually need to define your own `LatticeIter` type. You could just delegate to the iterator of `universes`, i.e.
```C++
{ return this->universes.begin(); }
```
And the same thing for `end`, `rbegin`, and `rend`.

Reply:
It's handy to have `LatticeIter` with the public `indx` attribute for distribcell purposes.  That `indx` value is used for indexing into `Lattice::offsets`

Reply:
I see, I see